stdlib/random.ml

```diff
@@ -176,6 +176,22 @@ module State = struct
 
   let bool s = (bits s land 1 = 0)
 
+  let bits32 s =
+    let b1 = Int32.(shift_right_logical (of_int (bits s)) 14) in  (* 16 bits *)
+    let b2 = Int32.(shift_right_logical (of_int (bits s)) 14) in  (* 16 bits *)
+    Int32.(logor b1 (shift_left b2 16))
+
+  let bits64 s =
+    let b1 = Int64.(shift_right_logical (of_int (bits s)) 9) in  (* 21 bits *)
+    let b2 = Int64.(shift_right_logical (of_int (bits s)) 9) in  (* 21 bits *)
+    let b3 = Int64.(shift_right_logical (of_int (bits s)) 8) in  (* 22 bits *)
+    Int64.(logor b1 (logor (shift_left b2 21) (shift_left b3 42)))
```

Comment:
Out of curiosity: other splits are possible (for example 30, 30, 4), and in particular using 30 bits in some cases could avoid some (probably neglectible) shift-right-logical operations. The current split uses as few "lower bits" from `bits` as possible. Is there a particular reason for this choice?

Reply:
There's a rumor that among the 30 random bits returned by `bits ()`, the high bits are "more random" than the low bits.  I don't have my TAOCP vol 2 with me to check.  But that's why the top 21/22 bits are used here: if you draw three sets of 30 random bits, use the "more random" bits of each set.

Reply:
Out of curiosity, I looked it up:

![image](https://user-images.githubusercontent.com/113560/126128938-4d599f2e-9f2c-4e91-aa57-b98d2bee2cc6.png)

In the book the analysis is done for a standard linear congruence generator, but the same argument applies to any generator using a power-of-two modulus, such as `2^30` in the Fibonacci generator used in `Random`.

Reply:
> The `(curval lxor ((curval lsr 25) land 0x1F))` in `Random` seems to take some high-order bits to the low-order bits. Does that counter in part this problem?

Very likely.  But that's @damiendoligez 's wizardry, so I'll let him comment.

Reply:
Also see the comment at top of `stdlib/random.ml`.

Reply:
> The `(curval lxor ((curval lsr 25) land 0x1F))` in `Random` seems to take some high-order bits to the low-order bits. Does that counter in part this problem?

I'm not Knuth and I didn't try to do the theory, but experimenting with DieHarder seemed to show that this is indeed better.  IIRC one or two of DieHarder tests did fail on sequences of the one or two low-order bits of the generator before this patch. Even with the patch, I approve of favoring the high-order bits.